python/whylogs/api/logger/experimental/logger/actor/process_rolling_logger_messages.py

```diff
@@ -0,0 +1,282 @@
+"""
+Messages that the process logger can process.
+
+These types are core types that you can send to the process logger. They are either subclasses of
+TypedDict or wrappers around serialized byte versions of those dicts. Everything in here is a
+TypedDict because we use orjson to deserialize them for performance reasons and that library prefers
+to output everything as dictionaries.
+
+The dataclass containers for those messages have helper methods to extract/deserialize the dicts. It's
+important to not raise exceptions in those data classes because the actor does a lot of large batch processing
+and exceptions would result in losing the entire batch. Instead, they return None and log errors if there is
+some issue deserializing or validating.
+"""
+import sys
+
+if sys.version_info >= (3, 8):
+    from typing import TypedDict  # pylint: disable=no-name-in-module
+else:
+    from typing_extensions import TypedDict
+
+import base64
+from dataclasses import dataclass, field
+from typing import Dict, List, Optional, Tuple, Union, cast
+
+from whylogs.api.logger.experimental.logger.actor.thread_rolling_logger import (
+    LoggerStatus,
+)
+
+try:
+    import orjson
+except ImportError:
+    raise ImportError("Install whylogs with extra [proc] for process based logging: pip install whylogs[proc]")
+import logging
+import uuid
+
+import numpy as np
+import pandas as pd
+
+from whylogs.api.logger.experimental.logger.actor.time_util import (
+    TimeGranularity,
+    truncate_time_ms,
+)
+
+DataTypes = Union[str, int, float, bool, List[float], List[int], List[str]]
+
+_logger = logging.getLogger("profile_actor_messages")
+
+
+class DataDict(TypedDict):
+    columns: List[str]
+    data: List[List[DataTypes]]
+
+
+def data_dict_from_pandas(df: pd.DataFrame) -> DataDict:
+    multiple = df.to_dict(orient="split")
+    del multiple["index"]  # get rid of this to min/max payload size
+    return {
+        "columns": multiple["columns"],
+        "data": multiple["data"],
+    }
+
+
+class LogRequestDict(TypedDict):
+    datasetId: str
+    timestamp: Optional[int]
+    multiple: DataDict
+
+
+class LogEmbeddingRequestDict(TypedDict):
+    datasetId: str
+    timestamp: Optional[int]
+    embeddings: Dict[str, List[DataTypes]]
+
+
+class PubSubMessage(TypedDict):
+    attributes: Dict[str, str]
+    data: str
+    message_id: str
+    publish_time: str
+
+
+class PubSubDict(TypedDict):
+    subscription: str
+    message: PubSubMessage
+    log_request: LogRequestDict
+
+
+class DebugMessage:
+    pass
+
+
+class FlushMessage:
+    pass
+
+
+@dataclass
+class ProcessLoggerStatus:
+    dataset_id: str
+    status: LoggerStatus
+
+
+@dataclass
+class ProcessLoggerStatusMessage:
+    id: str = field(default_factory=lambda: str(uuid.uuid4()))
+
```

Comment:
I'm not sure I follow the semantics of the different message types and how they relate, can you explain this one a bit?

Reply:
The message types are what determine what happens once they're delivered to the actor. The actor checks the message's type and then does _something_ with it. In the case of the `ProcessLoggerStatusMessage`, the actor will lookup the current internal state. For logging, the RawLogMessage has all of the data that has to be logged and the actor will use it to do actual logging.

They're mostly details of message passing. The user doesn't have to care about these. They can interract with the logger though the DataLogger interface which does the heavy lifting of sending the message and coordinating across processes. The process actor is both an Actor and a DataLogger, so we can expose it casted as a DataLogger if you want to hide all of this stuff. And its an actor of  course, so it processes messages serially from its queue.